hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/StreamBlockInputStream.java

```diff
@@ -207,36 +219,61 @@ protected synchronized void acquireClient() throws IOException {
     checkOpen();
     if (xceiverClient == null) {
       final Pipeline pipeline = pipelineRef.get();
+      final XceiverClientSpi client;
       try {
-        xceiverClient = xceiverClientFactory.acquireClientForReadData(pipeline);
+        client = xceiverClientFactory.acquireClientForReadData(pipeline);
       } catch (IOException ioe) {
         LOG.warn("Failed to acquire client for pipeline {}, block {}", pipeline, blockID);
         throw ioe;
       }
+
+      if (client == null) {
+        throw new IOException("Failed to acquire client for " + pipeline);
+      }
+      if (!(client instanceof XceiverClientGrpc)) {
+        throw new IOException("Unexpected client class: " + client.getClass().getName() + ", " + pipeline);
+      }
+
+      xceiverClient =  (XceiverClientGrpc) client;
     }
   }
 
-  private void initialize() throws IOException {
-    if (initialized) {
-      return;
-    }
-    while (true) {
+  private synchronized void initialize() throws IOException {
+    while (streamingReader == null) {
       try {
         acquireClient();
-        streamingReader = new StreamingReader();
-        ContainerProtocolCalls.readBlock(xceiverClient, position, blockID, tokenRef.get(),
-            pipelineRef.get().getReplicaIndexes(), streamingReader);
-        initialized = true;
-        return;
-      } catch (InterruptedException ie) {
-        Thread.currentThread().interrupt();
-        handleExceptions(new IOException("Interrupted", ie));
       } catch (IOException ioe) {
         handleExceptions(ioe);
       }
+
+      streamingReader = new StreamingReader();
+      xceiverClient.initStreamRead(blockID, streamingReader);
```

Comment:
Thanks @szetszwo for the patch. Should we clean up `streamingReader` somewhere? Otherwise, when we throw an exception here, we might never be able to `initialize` it unless we close the entire stream.

Reply:
@chungen0126 , there are nothing to clean up for `StreamingReader`:
```java
//StreamingReader
    private final BlockingQueue<ReadBlockResponseProto> responseQueue = new LinkedBlockingQueue<>();

    private final CompletableFuture<Void> future = new CompletableFuture<>();
    private final AtomicBoolean semaphoreReleased = new AtomicBoolean(false);
    private final AtomicReference<StreamingReadResponse> response = new AtomicReference<>();
```

Reply:
Yes, but if the `StreamingReader` is not null, it won't enter the while loop which spans the `initialize` method.